File: packages/android/formbricksSDK/src/main/java/com/formbricks/formbrickssdk/model/environment/Segment.kt
package com.formbricks.formbrickssdk.model.environment

import kotlinx.serialization.*
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.*
import kotlinx.serialization.descriptors.*
import kotlinx.serialization.encoding.*

// MARK: - Connector
@Serializable
enum class SegmentConnector {
    @SerialName("and") AND,
    @SerialName("or") OR
}

// MARK: - Filter Operators
@Serializable
enum class FilterOperator {
    @SerialName("lessThan") LESS_THAN,
    @SerialName("lessEqual") LESS_EQUAL,
    @SerialName("greaterThan") GREATER_THAN,
    @SerialName("greaterEqual") GREATER_EQUAL,
    @SerialName("equals") EQUALS,
    @SerialName("notEquals") NOT_EQUALS,
    @SerialName("contains") CONTAINS,
    @SerialName("doesNotContain") DOES_NOT_CONTAIN,
    @SerialName("startsWith") STARTS_WITH,
    @SerialName("endsWith") ENDS_WITH,
    @SerialName("isSet") IS_SET,
    @SerialName("isNotSet") IS_NOT_SET,
    @SerialName("userIsIn") USER_IS_IN,
    @SerialName("userIsNotIn") USER_IS_NOT_IN
}

// MARK: - Filter Value
@Serializable(with = SegmentFilterValueSerializer::class)
sealed class SegmentFilterValue {
    data class StringValue(val value: String) : SegmentFilterValue()
    data class NumberValue(val value: Double) : SegmentFilterValue()
}

object SegmentFilterValueSerializer : KSerializer<SegmentFilterValue> {
    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("SegmentFilterValue", PrimitiveKind.STRING)
    override fun deserialize(decoder: Decoder): SegmentFilterValue {
        val jsonInput = decoder as JsonDecoder
        val element = jsonInput.decodeJsonElement()
        return when (element) {
            is JsonPrimitive -> {
                element.doubleOrNull?.let { SegmentFilterValue.NumberValue(it) }
                    ?: SegmentFilterValue.StringValue(element.content)
            }
            else -> throw SerializationException("Unexpected type for SegmentFilterValue: $element")
        }
    }
    override fun serialize(encoder: Encoder, value: SegmentFilterValue) {
        val jsonOutput = encoder as JsonEncoder
        val element = when (value) {
            is SegmentFilterValue.NumberValue -> JsonPrimitive(value.value)
            is SegmentFilterValue.StringValue -> JsonPrimitive(value.value)
        }
        jsonOutput.encodeJsonElement(element)
    }
}

// MARK: - Filter Root
@Serializable(with = SegmentFilterRootSerializer::class)
sealed class SegmentFilterRoot {
    data class Attribute(val contactAttributeKey: String) : SegmentFilterRoot()
    data class Person(val personIdentifier: String) : SegmentFilterRoot()
    data class Segment(val segmentId: String) : SegmentFilterRoot()
    data class Device(val deviceType: String) : SegmentFilterRoot()
}

object SegmentFilterRootSerializer : KSerializer<SegmentFilterRoot> {
    override val descriptor: SerialDescriptor = buildClassSerialDescriptor("SegmentFilterRoot") {
        element<String>("type")
    }
    override fun deserialize(decoder: Decoder): SegmentFilterRoot {
        val input = decoder as JsonDecoder
        val obj = input.decodeJsonElement().jsonObject
        return when (val type = obj["type"]?.jsonPrimitive?.content) {
            "attribute" -> SegmentFilterRoot.Attribute(obj["contactAttributeKey"]!!.jsonPrimitive.content)
            "person"    -> SegmentFilterRoot.Person(obj["personIdentifier"]!!.jsonPrimitive.content)
            "segment"   -> SegmentFilterRoot.Segment(obj["segmentId"]!!.jsonPrimitive.content)
            "device"    -> SegmentFilterRoot.Device(obj["deviceType"]!!.jsonPrimitive.content)
            else         -> throw SerializationException("Unknown root type: $type")
        }
    }
    override fun serialize(encoder: Encoder, value: SegmentFilterRoot) {
        val output = encoder as JsonEncoder
        val json = buildJsonObject {
            when (value) {
                is SegmentFilterRoot.Attribute -> {
                    put("type", JsonPrimitive("attribute"))
                    put("contactAttributeKey", JsonPrimitive(value.contactAttributeKey))
                }
                is SegmentFilterRoot.Person -> {
                    put("type", JsonPrimitive("person"))
                    put("personIdentifier", JsonPrimitive(value.personIdentifier))
                }
                is SegmentFilterRoot.Segment -> {
                    put("type", JsonPrimitive("segment"))
                    put("segmentId", JsonPrimitive(value.segmentId))
                }
                is SegmentFilterRoot.Device -> {
                    put("type", JsonPrimitive("device"))
                    put("deviceType", JsonPrimitive(value.deviceType))
                }
            }
        }
        output.encodeJsonElement(json)
    }
}

// MARK: - Qualifier
@Serializable
data class SegmentFilterQualifier(
    @SerialName("operator") val `operator`: FilterOperator
)

// MARK: - Primitive Filter
@Serializable
data class SegmentPrimitiveFilter(
    val id: String,
    val root: SegmentFilterRoot,
    val value: SegmentFilterValue,
    val qualifier: SegmentFilterQualifier
)

// MARK: - Recursive Resource
@Serializable(with = SegmentFilterResourceSerializer::class)
sealed class SegmentFilterResource {
    data class Primitive(val filter: SegmentPrimitiveFilter) : SegmentFilterResource()
    data class Group(val filters: List<SegmentFilter>) : SegmentFilterResource()
}

object SegmentFilterResourceSerializer : KSerializer<SegmentFilterResource> {
    override val descriptor = buildClassSerialDescriptor("SegmentFilterResource") {
        // You can declare children here if you like,
        // or leave it empty if you’re purely passing through.
    }
    override fun deserialize(decoder: Decoder): SegmentFilterResource {
        val input = decoder as JsonDecoder
        val element = input.decodeJsonElement()
        return if (element is JsonArray) {
            val list = element.map { input.json.decodeFromJsonElement(SegmentFilter.serializer(), it) }
            SegmentFilterResource.Group(list)
        } else {
            val prim = input.json.decodeFromJsonElement(SegmentPrimitiveFilter.serializer(), element)
            SegmentFilterResource.Primitive(prim)
        }
    }
    override fun serialize(encoder: Encoder, value: SegmentFilterResource) {
        val output = encoder as JsonEncoder
        val json = when (value) {
            is SegmentFilterResource.Primitive -> output.json.encodeToJsonElement(SegmentPrimitiveFilter.serializer(), value.filter)
            is SegmentFilterResource.Group     -> output.json.encodeToJsonElement(ListSerializer(SegmentFilter.serializer()), value.filters)
        }
        output.encodeJsonElement(json)
    }
}

// MARK: - Filter Node
@Serializable
data class SegmentFilter(
    val id: String,
    val connector: SegmentConnector? = null,
    val resource: SegmentFilterResource
)

// MARK: - Segment Model
@Serializable
data class Segment(
    val id: String,
    val title: String,
    val description: String? = null,
    @SerialName("isPrivate") val isPrivate: Boolean,
    val filters: List<SegmentFilter>,
    val environmentId: String,
    val createdAt: String,
    val updatedAt: String,
    val surveys: List<String>
)

File: packages/ios/FormbricksSDK/FormbricksSDK/Model/Environment/Surveys/Segment.swift
import Foundation

// MARK: - Connector

enum SegmentConnector: String, Codable {
    case and
    case or
}

// MARK: - Filter Operators

/// Combined operator set for all filter types
enum FilterOperator: String, Codable {
    // Base / Arithmetic
    case lessThan
    case lessEqual
    case greaterThan
    case greaterEqual
    case equals
    case notEquals
    // Attribute / String
    case contains
    case doesNotContain
    case startsWith
    case endsWith
    // Existence
    case isSet
    case isNotSet
    // Segment membership
    case userIsIn
    case userIsNotIn
}

// MARK: - Filter Value

enum SegmentFilterValue: Codable {
    case string(String)
    case number(Double)

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let num = try? container.decode(Double.self) {
            self = .number(num)
        } else if let str = try? container.decode(String.self) {
            self = .string(str)
        } else {
            throw DecodingError.typeMismatch(
                SegmentFilterValue.self,
                DecodingError.Context(
                    codingPath: decoder.codingPath,
                    debugDescription: "Value is neither Double nor String"
                )
            )
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .number(let num):
            try container.encode(num)
        case .string(let str):
            try container.encode(str)
        }
    }
}

// MARK: - Root

enum SegmentFilterRoot: Codable {
    case attribute(contactAttributeKey: String)
    case person(personIdentifier: String)
    case segment(segmentId: String)
    case device(deviceType: String)

    private enum CodingKeys: String, CodingKey {
        case type
        case contactAttributeKey
        case personIdentifier
        case segmentId
        case deviceType
    }

    private enum RootType: String, Codable {
        case attribute
        case person
        case segment
        case device
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let type = try container.decode(RootType.self, forKey: .type)
        switch type {
        case .attribute:
            let key = try container.decode(String.self, forKey: .contactAttributeKey)
            self = .attribute(contactAttributeKey: key)
        case .person:
            let id = try container.decode(String.self, forKey: .personIdentifier)
            self = .person(personIdentifier: id)
        case .segment:
            let id = try container.decode(String.self, forKey: .segmentId)
            self = .segment(segmentId: id)
        case .device:
            let type = try container.decode(String.self, forKey: .deviceType)
            self = .device(deviceType: type)
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .attribute(let key):
            try container.encode(RootType.attribute, forKey: .type)
            try container.encode(key, forKey: .contactAttributeKey)
        case .person(let id):
            try container.encode(RootType.person, forKey: .type)
            try container.encode(id, forKey: .personIdentifier)
        case .segment(let id):
            try container.encode(RootType.segment, forKey: .type)
            try container.encode(id, forKey: .segmentId)
        case .device(let type):
            try container.encode(RootType.device, forKey: .type)
            try container.encode(type, forKey: .deviceType)
        }
    }
}

// MARK: - Qualifier

struct SegmentFilterQualifier: Codable {
    let `operator`: FilterOperator
}

// MARK: - Primitive Filter

struct SegmentPrimitiveFilter: Codable {
    let id: String
    let root: SegmentFilterRoot
    let value: SegmentFilterValue
    let qualifier: SegmentFilterQualifier

    // Add run-time refinements if needed
}

// MARK: - Recursive Filter Resource

enum SegmentFilterResource: Codable {
    case primitive(SegmentPrimitiveFilter)
    case group([SegmentFilter])

    init(from decoder: Decoder) throws {
        // Try primitive first
        if let prim = try? SegmentPrimitiveFilter(from: decoder) {
            self = .primitive(prim)
        } else {
            let nested = try [SegmentFilter](from: decoder)
            self = .group(nested)
        }
    }

    func encode(to encoder: Encoder) throws {
        switch self {
        case .primitive(let prim):
            try prim.encode(to: encoder)
        case .group(let arr):
            try arr.encode(to: encoder)
        }
    }
}

// MARK: - Base Filter (node)

struct SegmentFilter: Codable {
    let id: String
    let connector: SegmentConnector?
    let resource: SegmentFilterResource
}

// MARK: - Segment Model

struct Segment: Codable {
    let id: String
    let title: String
    let description: String?
    let isPrivate: Bool
    let filters: [SegmentFilter]
    let environmentId: String
    let createdAt: Date
    let updatedAt: Date
    let surveys: [String]

    private enum CodingKeys: String, CodingKey {
        case id, title, description, filters, surveys
        case isPrivate = "isPrivate"
        case environmentId, createdAt, updatedAt
    }
}
